File: leetcode/2300.js
/**
 * @param {number[]} spells
 * @param {number[]} potions
 * @param {number} success
 * @return {number[]}
 */

/**
 * Binary search to find the first index where potions[index] * spell >= success
 * @param {number[]} potions
 * @param {number} spell
 * @param {number} success
 * @return {number}
 */
const binarySearch = (potions, spell, success) => {
  let left = 0;
  let right = potions.length - 1;

  while (left <= right) {
    const mid = Math.floor((left + right) / 2);
    if (potions[mid] * spell >= success) {
      right = mid - 1;
    } else {
      left = mid + 1;
    }
  }

  return left;
};

/**
 * @param {number[]} spells
 * @param {number[]} potions
 * @param {number} success
 * @return {number[]}
 */
const successfulPairs = (spells, potions, success) => {
  const sortedPotions = potions.slice().sort((a, b) => a - b);

  return spells.map(spell => {
    const firstSuccessfulIndex = binarySearch(sortedPotions, spell, success);
    return sortedPotions.length - firstSuccessfulIndex;
  });
};

File: leetcode/2300.py
class Solution:
    def successfulPairs(self, spells: List[int], potions: List[int], success: int) -> List[int]:
        sorted_potions: int = sorted(potions)
        num_potions: int = len(potions)

        def find_potions_ind(spell: int) -> int:
            left: int = 0
            right: int = num_potions - 1

            while left <= right:
                mid: int = floor((left + right) / 2)
                val: int = sorted_potions[mid]

                if (val * spell) >= success:
                    right = mid - 1
                else:
                    left = mid + 1

                return left

        return list(map(lambda s: num_potions - find_potions_ind(s), spells))
